src/staticdata.c

```diff
@@ -1563,38 +1573,83 @@ static jl_value_t *strip_codeinfo(jl_method_t *m, jl_value_t *ci_, int isdef)
-            jl_gc_wb(codeinst, codeinst->inferred);
+            if (jl_options.strip_ir) {
+                record_field_change(&codeinst->inferred, jl_nothing);
+            }
+            else if (jl_options.strip_metadata) {
+                codeinst->inferred = strip_codeinfo_meta(mi->def.method, codeinst->inferred, 0);
+                jl_gc_wb(codeinst, codeinst->inferred);
+            }
         }
         codeinst = jl_atomic_load_relaxed(&codeinst->next);
     }
+    if (jl_options.strip_ir) {
+        record_field_change(&mi->uninferred, NULL);
+        record_field_change((jl_value_t**)&mi->backedges, NULL);
+        record_field_change((jl_value_t**)&mi->callbacks, NULL);
+        record_field_change(&mi->specTypes, (jl_value_t*)jl_emptytuple_type);
+    }
 }
 
 static int strip_all_codeinfos__(jl_typemap_entry_t *def, void *_env)
 {
     jl_method_t *m = def->func.method;
     if (m->source) {
-        m->source = strip_codeinfo(m, m->source, 1);
-        jl_gc_wb(m, m->source);
+        int stripped_ir = 0;
+        if (jl_options.strip_ir) {
+            if (m->unspecialized) {
+                jl_code_instance_t *unspec = jl_atomic_load_relaxed(&m->unspecialized->cache);
+                if (unspec && jl_atomic_load_relaxed(&unspec->invoke)) {
+                    // we have a generic compiled version, so can remove the IR
+                    record_field_change(&m->source, jl_nothing);
+                    stripped_ir = 1;
+                }
+            }
+            if (!stripped_ir) {
+                int mod_setting = jl_get_module_compile(m->module);
+                // if the method is declared not to be compiled, keep IR for interpreter
+                if (!(mod_setting == JL_OPTIONS_COMPILE_OFF || mod_setting == JL_OPTIONS_COMPILE_MIN)) {
+                    record_field_change(&m->source, jl_nothing);
+                    stripped_ir = 1;
+                }
+            }
+        }
+        if (jl_options.strip_metadata && !stripped_ir) {
+            m->source = strip_codeinfo_meta(m, m->source, 1);
+            jl_gc_wb(m, m->source);
+        }
     }
-    jl_svec_t *specializations = def->func.method->specializations;
+    jl_svec_t *specializations = m->specializations;
     size_t i, l = jl_svec_len(specializations);
     for (i = 0; i < l; i++) {
         jl_value_t *mi = jl_svecref(specializations, i);
         if (mi != jl_nothing)
             strip_specializations_((jl_method_instance_t*)mi);
     }
+    if (jl_options.strip_ir) {
+        record_field_change((jl_value_t**)&m->specializations, (jl_value_t*)jl_emptysvec);
+        record_field_change((jl_value_t**)&m->speckeyset, jl_an_empty_vec_any);
+    }
+    if (m->unspecialized)
+        strip_specializations_(m->unspecialized);
     return 1;
 }
 
 static void strip_all_codeinfos_(jl_methtable_t *mt, void *_env)
 {
+    if (jl_options.strip_ir)
+        record_field_change((jl_value_t**)&mt->backedges, NULL);
```

Comment:
world-ages aren't just part of the IR, it is part of the generated code and return type
```suggestion
```